### PR TITLE
add WeakValueCache and fix WeakValueDict

### PR DESCRIPTION
**Pull request by tthsqe12**
I am not interested ATM in making a WeakValueDict, partly because it has many superfluous requirements, partly because I don't like locks in single-threaded code, and mainly because I don't see how it can coherently fulfill the dictionary interface.
I have not observed the AA or Nemo test code failing with this WeakValueCache, but to be fair, I did not observed the WeakValueDict failing on the same machine.

### CODE REVIEW

**Comment by tthsqe12** [comment]
There is no work in progress here. The weakvalue{dict|cache} is just dead code that someone might want to play with at some point in the future. I don't think it can actually work reliably until julia gets better.

**Comment by tthsqe12** [comment]
Turns out Julia got better at 1.6 :)

**Comment by thofma** [comment]
Can you add a test?

**Comment by tthsqe12** [comment]
tests might come next week

**Comment by fingolfin** [comment]
I don't understand, what exactly is wrong with `WeakValueDict` in Julia 1.6? 

As far as I can tell `WeakValueCache` differs mainly in that it doesn't claim to be a full Dict and thus does not support some things that are difficult or impossible to implement in the weak value context. Is that it? Which operations specifically were/are problematic with WVD, though?

We really would like to use WVD (or perhaps WVC, if it can do the job) in Oscar.jl in the groups code. We require Julia >= 1.6  in Oscar.jl, Singular.jl, Polymake.jl, GAP.jl already  so we don't have to worry about problems with older Julia there. Of course AA and Nemo still (cl)aim to support older Julia versions (although I hope this will change soon?). But for now we could simply only include the WVC/WVD code if Julia >= 1.6. Or we move it to a separate package? But I really need some functionality like this.

I'd also be happy to help debug it but, so far I only heard murmurs and claims about problems that could be seen by waiting long enough for some CI tests to fail, but that's really vague...?

File: src/WeakValueDict.jl [hunk 1]
```diff
@@ -145,26 +511,34 @@ function Base.get(wkh::WeakValueDict{K, V}, key, default) where {K, V}
...
     lock(wkh) do
-        return pop!(wkh.ht, key).value
+        return pop!(wkh.ht, key).value  # ???
     end
 end
 function Base.pop!(wkh::WeakValueDict{K}, key, default) where {K}
     key === nothing && return default
     lock(wkh) do
-        return pop!(wkh.ht, key, default)
+        return pop!(wkh.ht, key, default)  # ect.
```
Comment (fingolfin) [inline src/WeakValueDict.jl:541]: Do you mean "etc." ? But what is this comment supposed to tell me?

File: src/WeakValueDict.jl [hunk 2]
```diff
@@ -145,26 +511,34 @@ function Base.get(wkh::WeakValueDict{K, V}, key, default) where {K, V}
...
     lock(wkh) do
-        return get(default, wkh.ht, key)
+        return get(default, wkh.ht, key)  # ???
     end
 end
 function Base.pop!(wkh::WeakValueDict{K}, key) where {K}
     key === nothing && throw(KeyError(nothing))
     lock(wkh) do
-        return pop!(wkh.ht, key).value
+        return pop!(wkh.ht, key).value  # ???
```
Comment (fingolfin) [inline src/WeakValueDict.jl:535]: ?????

File: src/WeakValueDict.jl [hunk 3]
```diff
@@ -145,26 +511,34 @@ function Base.get(wkh::WeakValueDict{K, V}, key, default) where {K, V}
...
     end
 end
+
+#### the rest of these look dodgy
+
 function Base.get(default::Base.Callable, wkh::WeakValueDict{K}, key) where {K}
     key === nothing && throw(KeyError(nothing))
     lock(wkh) do
-        return get(default, wkh.ht, key)
+        return get(default, wkh.ht, key)  # ???
```
Comment (fingolfin) [inline src/WeakValueDict.jl:529]: Could you elaborate?

File: src/WeakValueDict.jl [hunk 4]
```diff
@@ -1,9 +1,364 @@
...
+   _setindex!(h, WeakRef(x), key, abs(index))
+   return h
+end
+
+
+################################################################################
+#
+# WeakValueDict
+# This section is a modification of base/weakkeydict.jl from the Julia project.
+# It pretends to satisfy the dictionary interface.
```
Comment (fingolfin) [inline src/WeakValueDict.jl:356]: This is not helpful if you don't also say in how far this is a pretense and how it breaks down exactly?

File: src/WeakValueDict.jl [hunk 5]
```diff
@@ -1,9 +1,364 @@
...
+
+### setindex! ####
+
+function setindex!(h::WeakValueCache{K, V}, v0, key0) where {K, V}
+   key = convert(K, key0)
+   isequal(key, key0) || throw(ArgumentError("$key0 is not a valid key for type $K"))
+   setindex!(h, v0, key)
+end
+
+# the WeakRef makes key conversion essentially useless
```
Comment (fingolfin) [inline src/WeakValueDict.jl:343]: Do you mean value conversion (not key conversion)? 

I agree that one probably only wants to store the exact object one passed in, and this implicit conversion, will just hide accidental conversions / inefficiencies.

Although strictly speaking the result won't necessarily be useless... but if someone wants this, they might as well insert an *explicit* conversion instead.

File: src/WeakValueDict.jl [hunk 6]
```diff
@@ -1,9 +1,364 @@
...
+   if index > 0
+      x = h.vals[index].value
+      if x !== nothing
+         return x::V
+      end
+   else
+      index = -index
+   end
+   age0 = h.age
+   x = convert(V, default())
```
Comment (fingolfin) [inline src/WeakValueDict.jl:325]: perhaps drop this for the same reason as in `setindex!`?
```suggestion
   x = default()::V
```

**Review by fingolfin — CHANGES_REQUESTED** [review]

File: src/WeakValueDict.jl (hunk 5 above)
Comment (tthsqe12) [reply src/WeakValueDict.jl:343]: oops, I added this comment by mistake when I was not thinking clearly. Key conversions are clearly OK in a weak*value*dict/cache.

File: src/WeakValueDict.jl (hunk 2 above)
Comment (tthsqe12) [reply src/WeakValueDict.jl:535]: Is it ok for `pop!` to return nothing? Do we keep `pop!`'ing until we the poped value is not nothing or we can't `pop!` anymore? These are the kind of questions imposed by fake dictionaries...

File: src/WeakValueDict.jl (hunk 4 above)
Comment (tthsqe12) [reply src/WeakValueDict.jl:356]: will try

File: src/WeakValueDict.jl (hunk 6 above)
Comment (tthsqe12) [reply src/WeakValueDict.jl:325]: ok. I understand.

**Comment by tthsqe12** [comment]
It looks like many of @fingolfin 's comments are directed towards my critisisms of `WeakValueDict{K,V} <: AbstractDict{K,V}`. I will try to improve the `WeakValueDict` code and put better comments in. For now, I can say
0. Where is the list of methods an `AbstractDict` is supposed to implement?
1. Is the one-argument `pop!(collection)` also supposed to work for wvd? What should it do?
2. The most serious problem I can see with wvd is `length`. The iterator for wvd is not guaranteed to have the same length as a previous call to `length`. How big of a problem is this for julia?

File: src/WeakValueDict.jl (hunk 1 above)
Comment (fingolfin) [reply src/WeakValueDict.jl:541]: So how about
```suggestion
        return pop!(wkh.ht, key, default)  # FIXME: this returns a WeakRef, but shouldn't
```
or
```suggestion
        return pop!(wkh.ht, key, default).value
```
And if `nothing` is still possible at this stage, then of course that needs to be checked for and `default` returned.

File: src/WeakValueDict.jl (hunk 2 above)
Comment (fingolfin) [reply src/WeakValueDict.jl:535]: If the value is nothing at this point, then this is equivalent to "the mapping does not exist", in which case the documentation says
> throw an error if default is not specified

**Comment by thofma** [comment]
I think the biggest problem (which @tthsqe12 already mentioned) is, that there is no `AbstractDict` interface (https://github.com/JuliaLang/julia/issues/25941).

From what I understand we want to provide the following subset of functions:
- [ ] `get/get!/set!`
- [ ] `empty!`
- [ ] `length` (useful for introspection, no serious applications)
- [ ] `delete!`
- [ ] `pop!` (because we already have it, just needs an error)

If we agree then @tthsqe12 can tick of the ones we already have and maybe add the ones we want.

This will be our local interpretation of the "`AbstractDict` interface" and IMHO we should keep it `<: AbstractDict`.

**Comment by thofma** [comment]
Sorry for the probably stupid question, but does this PR contain both `WeakValueDict` and `WeakValueCache`?

**Review by fingolfin — APPROVED** [review]
Sorry, didn't see that you were waiting for me.

I haven't reviewed the code in this PR in depth, but in principle, I have no objections. So feel free to go ahead and merge it.

**Pull request merged**